# Проект по SQL

## Цель исследования
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.


Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных

### Таблица `books`
Содержит данные о книгах:

    `book_id` — идентификатор книги;
    `author_id` — идентификатор автора;
    `title` — название книги;
    `num_pages` — количество страниц;
    `publication_date` — дата публикации книги;
    `publisher_id` — идентификатор издателя.
    
    
### Таблица authors
Содержит данные об авторах:

    `author_id` — идентификатор автора;
    `author` — имя автора.


### Таблица publishers
Содержит данные об издательствах:
    
    `publisher_id` — идентификатор издательства;
    `publisher` — название издательства;

### Таблица ratings
Содержит данные о пользовательских оценках книг:

    `rating_id` — идентификатор оценки;
    `book_id` — идентификатор книги;
    `username` — имя пользователя, оставившего оценку;
    `rating` — оценка книги.

### Таблица reviews
Содержит данные о пользовательских обзорах:

    `review_id` — идентификатор обзора;
    `book_id` — идентификатор книги;
    `username` — имя автора обзора;
    `text` — текст обзора.


## Задания

* Посчитайте, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитайте количество обзоров и среднюю оценку;
* Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
* Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
* Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = ''' SELECT * FROM {}; 
    '''.format(table) 

    request = pd.io.sql.read_sql(query, con = engine) 
# смотрим таблицы
    display(request.head())
    print(request.info())
    print(f'Таблица: "{table}"')
    print('------------------------------------------------------')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
Таблица: "books"
------------------------------------------------------


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
Таблица: "authors"
------------------------------------------------------


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
Таблица: "publishers"
------------------------------------------------------


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
Таблица: "ratings"
------------------------------------------------------


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
Таблица: "reviews"
------------------------------------------------------


### Предварительный вывод:
* Данные нормальные, пропусков нет. 


In [3]:
# запишем каждую таблицу в датафрейм
query_books = ''' SELECT * FROM books; ''' 
books = pd.io.sql.read_sql(query_books, con = engine)


query_reviews = ''' SELECT * FROM reviews; ''' 
reviews = pd.io.sql.read_sql(query_reviews, con = engine)


query_ratings = ''' SELECT * FROM ratings; ''' 
ratings = pd.io.sql.read_sql(query_ratings, con = engine)


query_publishers = ''' SELECT * FROM publishers; ''' 
publishers = pd.io.sql.read_sql(query_publishers, con = engine)

query_authors = ''' SELECT * FROM authors; ''' 
authors = pd.io.sql.read_sql(query_authors, con = engine)

## Считаем - сколько книг вышло после 1 января 2000 года:

In [4]:
query_1 = ''' 
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-1-1'; 
''' 

book = pd.io.sql.read_sql(query_1, con = engine)


print(f'После 01.01.2000 г. вышло: {book.iloc[0,0]} книг.')

После 01.01.2000 г. вышло: 819 книг.


### Предварительный вывод:
* После 01.01.2000 г. вышло: 819 книг

## Для каждой книги считаем кол-во обзоров и среднюю оценку:

In [5]:
query_2 = ''' 

SELECT 
    books.title AS book_name,
    books.book_id AS id,
    SubRev.n_rev AS reviews,
    SubRat.avg_rat AS avg_rating
FROM books LEFT JOIN 
                (SELECT
                    book_id,
                    AVG(rating) AS avg_rat
                FROM
                    ratings
                GROUP BY 
                    book_id
                ) AS SubRat ON SubRat.book_id = books.book_id
        LEFT JOIN
                (SELECT 
                    book_id,
                    COUNT(review_id) AS n_rev
                FROM
                    reviews
                GROUP BY 
                    book_id
                ) AS SubRev ON SubRev.book_id = books.book_id
                
; ''' 

book_review = pd.io.sql.read_sql(query_2, con = engine)
book_review

book_name   id  reviews  \
0            The Body in the Library (Miss Marple  #3)  652      2.0   
1                                            Galápagos  273      2.0   
2                             A Tree Grows in Brooklyn   51      5.0   
3    Undaunted Courage: The Pioneering First Missio...  951      2.0   
4                                          The Prophet  839      4.0   
..                                                 ...  ...      ...   
995                                Alice in Wonderland   64      4.0   
996          A Woman of Substance (Emma Harte Saga #1)   55      2.0   
997                                          Christine  148      3.0   
998  The Magicians' Guild (Black Magician Trilogy  #1)  790      2.0   
999                           The Plot Against America  828      2.0   

     avg_rating  
0      4.500000  
1      4.500000  
2      4.250000  
3      4.000000  
4      4.285714  
..          ...  
995    4.230769  
996    5.000000  
997    3.428571  
998    3.500000  
999    3.000000  

[1000 rows x 4 columns]

### Предварительный вывод:
Таблица с подсчётом на каждую книгу - кол-во обзоров и средня оценка готова

## Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры:

In [6]:
query_3 = ''' 
SELECT
    T.publisher_id,
    T.publisher,
    T.n_books
FROM
    (SELECT 
        books.publisher_id AS publisher_id,
        COUNT(books.book_id) AS n_books,
        publishers.publisher AS publisher 
    FROM 
        books 
        LEFT JOIN publishers ON books.publisher_id=publishers.publisher_id

    WHERE 
        num_pages>50
    GROUP BY
        books.publisher_id,
        publishers.publisher
    ORDER BY 
        n_books DESC) AS T
LIMIT 1
; ''' 

book_volume = pd.io.sql.read_sql(query_3, con = engine)
book_volume

publisher_id      publisher  n_books
0           212  Penguin Books       42

### Предварительный вывод:
`Penguin Books` - издательство, которое выпустило наибольшее число книг толще 50 страниц : 42 книги.

## Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками:

In [7]:
query_4 = ''' 

SELECT 
    authors.author AS author
   , AVG(ratings.rating) AS avg_rating
FROM books
    LEFT JOIN authors ON books.author_id = authors.author_id
    LEFT JOIN ratings ON books.book_id = ratings.book_id
GROUP BY 
     books.book_id
     , authors.author_id
HAVING
    COUNT(ratings.rating_id) >=50
ORDER BY 
    AVG(ratings.rating) DESC
LIMIT 5
    
; ''' 


author = pd.io.sql.read_sql(query_4, con = engine)
author

author  avg_rating
0         J.K. Rowling/Mary GrandPré    4.414634
1                     J.R.R. Tolkien    4.391892
2         J.K. Rowling/Mary GrandPré    4.287500
3  Markus Zusak/Cao Xuân Việt Khương    4.264151
4         J.K. Rowling/Mary GrandPré    4.246575

### Предварительный вывод:
Автор с самой высокой средней оценкой книг - `J.K. Rowling/Mary GrandPré`. Средний рейтинг 4,41

## Посчитаем среднее кол-во обзоров от пользователей, которые поставили больше 50 оценок:


In [8]:
query_5 = ''' 

SELECT
    AVG(Sub.n_reviews)
FROM
    (SELECT
        username,
        COUNT(review_id) AS n_reviews
    FROM
        reviews
    WHERE 
        username IN (
                SELECT 
                    username
                FROM 
                    ratings
                GROUP BY 
                    username
                HAVING 
                    COUNT(rating_id) > 50
                    )
    GROUP BY 
        username
    ) AS Sub
; ''' 

user_overview = pd.io.sql.read_sql(query_5, con = engine)
user_overview

avg
0  24.333333

### Предварительный вывод:
Среднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24,33


## Вывод исследования:

* После 01.01.2000 г. вышло: 819 книг
* Создали таблицу с подсчётом на каждую книгу - кол-во обзоров и средня оценка
* `Penguin Books` - издательство, которое выпустило наибольшее число книг толще 50 страниц : 42 книги.
* Автор с самой высокой средней оценкой книг - `J.K. Rowling/Mary GrandPré`. Средний рейтинг 4,41
* Среднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24,33